File: 백준/Gold/2473. 세 용액/세 용액.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static StringTokenizer st;

    public static void main(String[] args) throws Exception {
        int N = Integer.parseInt(br.readLine());
        long[] arr = new long[N];

        st = new StringTokenizer(br.readLine());
        Arrays.setAll(arr, i -> Long.parseLong(st.nextToken()));

        Arrays.sort(arr);
        long min = Long.MAX_VALUE;
        long[] result = new long[3];

        for (int i = 0; i < N - 2; i++) {
            int left = i + 1;
            int right = N - 1;

            while (left < right) {
                long sum = arr[i] + arr[left] + arr[right];

                if (Math.abs(sum) < min) {
                    min = Math.abs(sum);
                    result[0] = arr[i];
                    result[1] = arr[left];
                    result[2] = arr[right];
                }

                if (sum < 0) left++;
                else if (sum > 0) right--;
                else break;
            }
        }

        Arrays.sort(result);
        for (int i = 0; i < 3; i++) bw.write(result[i] + " ");
        bw.close();
    }
}

File: 백준/Gold/2473. 세 용액/세 용액.py
import math
import sys

input = lambda: sys.stdin.readline().rstrip()

if __name__ == '__main__':
    N = int(input())

    arr = list(map(int, input().split()))
    arr.sort()

    result = ()
    min_val = math.inf

    for i in range(N - 2):
        start = i + 1
        end = N - 1

        while start < end:
            left_val = arr[start]
            right_val = arr[end]

            total = right_val + left_val + arr[i]
            if min_val > abs(total):
                min_val = abs(total)
                result = arr[i], left_val, right_val
            # 용액이 양수라면
            if total < 0:
                start += 1
            else:
                end -= 1

    print(*result)